### ISSUE TITLE
[RFC]: Add control panel support for vLLM

### ISSUE THREAD

**Issue by leiwen83** [vllm-project/vllm#4873]
### Motivation.

Fastchat-vllm working model provide great benefit to us in deploying LLM in product serving. [1](https://blog.vllm.ai/2023/06/20/vllm.html)

The design of controller in Fastchat provide special benfit in llm serving, thanks for its loose decoupled design with vllm backend, so that,
* vllm backend could join in and leave the cluster freely, which provide autoscale capability
* With new model with different name join in, cluster could graduately update model in the cluster, also naming as roll update
* With central entrance point, user don't need to tag different url/ip to different ip, only send their request to the controller, controller would handle the rest, like dispatching request to proper backend according to model name, and handle load balance well.

The problem for fastchat is it need to handle many backend, beside vllm, so that seems to me it cannot keep up with the pace of evolving speed of vllm. It is very sad to see fastchat nowaday cannot support latest vllm feature like multi-lora, broken chat stream support, guidance decoding and so on...

Refence:
[1] https://blog.vllm.ai/2023/06/20/vllm.html

### Proposed Change.

So just head it up, I port the key feature of controller from fastchat, and make it at minimal shape, which for interface like /v1/../completions, it simply extract model name, and forward anything towards the backend, so that all feature of vllm could be used.

Current implement: #4861 

- [x] /v1/completions:  same interface of vllm's
- [x] /v1/chat/completions:  same interface of vllm's
- [x] /list_models: list models' name registered into controller
- [x] /health: check controller health status
- [x] /list_workers: list worker's detailed status, models provided by each worker, and its serving status
- [x] load balance with shortest queue algo
- [x] heart beat keep alive between controller and worker

Future directions:
- [ ] maybe rust could be used for reimplement the controller, if we find the performance could be improved a lot
- [ ] more algo for load balance
- [ ] unified metrics exposed by controller, which collected from each worker

### CC List.

@simon-mo @robertgshaw2-neuralmagic

**Comment by robertgshaw2-neuralmagic** [comment]
@leiwen83 

Can you explain why this type of control plane should be part of vLLM as opposed to something that Kubernetes should handle? Most of the functionality you list here are core features of KServe and it seems like it will be quite difficult for us to make a production grade system

**Comment by rkooo567** [comment]
+1 on the comment above. We should be very careful extending the scope of the project because many times it just becomes a suboptimal solution. 

I am curious about what's the main reason of "However, the challenge for Fastchat lies in managing multiple backends, including vLLM.", and just solving this problem is sufficient to make vllm compatible to fastchat?

**Comment by leiwen83** [comment]
@robertgshaw2-neuralmagic @rkooo567 
I fully understand your concern for the scope which may hurt vllm's production quality.

From the past few months use experience with fastchat, the mosting extracting feature is its ”Centralized Access“ in daily management, which mean unique url and models could be managed in one big pool. autoscale/load balance are the feature naturally with this central control panel.

I am not sure k8s/knative/kserve whether could accomplish this... From my knowledge, they could only do single modle serving? Certainly kserve could do autoscale/load balance well.

And the main reason why I brought up this topic, is I saw many issues and PR has been post in fastchat community regarding how to support latest vllm feature but no responsing later. So this is why I want to seek for help for vllm community's help for how to address ”Centralized Access“ requirement in our daily usage...

**Comment by njhill** [comment]
@leiwen83 I agree with other comments that this seems out of scope itself. From an autoscaling pov the important thing for vLLM itself is to expose appropriate metrics that existing cluster management systems (such as Kubernetes) can monitor and use to make and execute the scaling decisions.

The best metric probably isn't queue length, since the nature of requests in the queue(s) could be very different, and when queues are empty you would still want to balance based on the current state of requests being processed in each server.

In terms of centralized access to multiple models, this would just be an independent routing layer that can forward to vLLM deployments as appropriate. Of course the API that is exposed/forwarded should match the vLLM (OpenAI) API or forward payloads unchanged. If there are API parameters not propagated by the routing layer then it should be updated accordingly.

I do think that specialized load balancing is important though, to take into account the same queue-like metrics mentioned above, and to be session and/or prefix aware to maximize prefix cache hit-rates (so that consecutive requests which have the same chat history will hit the same server). @simon-mo mentioned that he was looking into this.

I'm not sure that python would necessarily be the best choice for such a router though.

**Comment by ishaan-jaff** [comment]
hi @leiwen83 @njhill I'm the [LiteLLM](https://github.com/BerriAI/litellm) maintainer - I believe we can be helpful here. 
(sorry if I misunderstood the problem on this issue - just trying to be helpful)

- You can use LiteLLM to Load balance between 100+ LLMs. We support the following routing strategies `"simple-shuffle", "least-busy", "usage-based-routing", "latency-based-routing", "cost-based-routing"`). Doc load balancing here: https://docs.litellm.ai/docs/proxy/reliability
- Use LiteLLM Virtual Keys to control model access, set budgets, rate limits https://docs.litellm.ai/docs/proxy/virtual_keys
- We support calling vLLM OpenAI compatible endpoints https://docs.litellm.ai/docs/providers/openai_compatible

**Comment by Jeffwan** [comment]
Let's remove the stable label. we have a solution plan to open source soon for the control plane part. This is still on track. /cc @simon-mo

**Comment by hmellor** [comment]
@Jeffwan how is this control plane progressing?

**Comment by Jeffwan** [comment]
@hmellor I am working with @simon-mo on the contribution process. We should have some updates earlier next week. it includes some routing, autoscaler, external kv cache,  dynamic lora allocation capabilities. Could you share your some details on what you are looking for?

**Comment by hmellor** [comment]
I'm not looking for anything in particular. I was just doing some housekeeping and noticed that this RFC hadn't been touched in a while. Thank you for the update!

**Comment by thousandhu** [comment]
> https://github.com/vllm-project/aibrix is open sourced and it's designed in cloud native way. It can play the role of vLLM control plane. I will help evaluate all features people proposed here are addressed. If not, I think we can consider to build on top of AIBrix.

I'm also nterested in AIBrix + vllm too. Please pin me when there are some progress. thank you
